fish-rust/src/wutil/wcstod.rs

```diff
@@ -0,0 +1,444 @@
+use super::errors::Error;
+use crate::wchar::IntoCharIter;
+use fast_float::parse_partial_iter;
+
+/// Parses a 64-bit floating point number.
+///
+/// Leading whitespace and trailing characters are ignored. If the input
+/// string does not contain a valid floating point number (where e.g.
+/// `"."` is seen as a valid floating point number), `None` is returned.
+/// Otherwise the parsed floating point number is returned.
+///
+/// The `decimal_sep` parameter is used to specify the decimal separator.
+/// '.' is a normal default.
+///
+/// The `consumed` parameter is used to return the number of characters
+/// consumed, similar to the "end" parameter to strtod.
+/// This is only meaningful if parsing succeeds.
+///
+/// Error::Overflow is returned if the value is too large in magnitude.
+pub fn wcstod<Chars>(input: Chars, decimal_sep: char, consumed: &mut usize) -> Result<f64, Error>
+where
+    Chars: IntoCharIter,
+{
+    let chars = input.chars();
+    if chars.clone().next().is_none() {
+        *consumed = 0;
+        return Err(Error::Empty);
+    }
+
+    let ret = parse_partial_iter(chars.clone().fuse(), decimal_sep);
```

Comment:
fastfloat does not support hexadecimal floats, try e.g. `printf %f 0x2`:

old: `2.000000`
new:
```
0x2: value not completely converted (can't convert 'x2')
0.000000
```

`hexf-parse` is used in `fast-float` tests for this, we should probably pull that in as well?

Reply:
Well shit. That's incredibly well spotted. I have to fix this.

Reply:
I've opted to fix this using (a forked, what else) version of [hexponent](https://docs.rs/hexponent/latest/hexponent/). I'm not happy about it since it's known incorrect for denormals but I don't know what else to do.